core/src/main/java/org/pentaho/di/core/plugins/PluginRegistry.java

```diff
@@ -0,0 +1,1114 @@
+   */
+  private PluginRegistry() {
+  }
+
+  /**
+   * @return The one and only PluginRegistry instance
+   */
+  public static PluginRegistry getInstance() {
+    return pluginRegistry;
+  }
+
+  private static Comparator<String> getNaturalCategoriesOrderComparator( Class<? extends PluginTypeInterface> pluginType ) {
+    PluginTypeCategoriesOrder naturalOrderAnnotation = pluginType.getAnnotation( PluginTypeCategoriesOrder.class );
+    final String[] naturalOrder;
+    if ( naturalOrderAnnotation != null ) {
+      String[] naturalOrderKeys = naturalOrderAnnotation.getNaturalCategoriesOrder();
+      Class<?> i18nClass = naturalOrderAnnotation.i18nPackageClass();
+
+      naturalOrder = Arrays.stream( naturalOrderKeys )
+        .map( key -> BaseMessages.getString( i18nClass, key ) )
+        .toArray( String[]::new );
+    } else {
+      naturalOrder = null;
+    }
+    return ( s1, s2 ) -> {
+      if ( naturalOrder != null ) {
+        int idx1 = Const.indexOfString( s1, naturalOrder );
+        int idx2 = Const.indexOfString( s2, naturalOrder );
+        // See PDI-17581, allows custom categories from plugins since TreeSet uses Comparator equality
+        if ( idx1 > -1 && idx2 > -1 ) {
+          // both items are in the natural ordering
+          return idx1 - idx2;
+        } else if ( idx1 > -1 ) {
+          // only the first item is in the natural ordering
+          return -1;
+        } else if ( idx2 > -1 ) {
+          // the second item is in the natural ordering
+          return 1;
+        }
+      }
+      // There is no natural ordering, or neither element being compared is in it
+      return String.CASE_INSENSITIVE_ORDER.compare( s1, s2 );
+    };
+  }
+
+  public void registerPluginType( Class<? extends PluginTypeInterface> pluginType ) {
+    lock.writeLock().lock();
+    try {
+      pluginMap.computeIfAbsent( pluginType, k -> new TreeSet<>( Plugin.nullStringComparator ) );
+
+      // Keep track of the categories separately for performance reasons...
+      //
+      categoryMap.computeIfAbsent( pluginType, k -> new TreeSet<>(
+        getNaturalCategoriesOrderComparator( pluginType ) ) );
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  public void removePlugin( Class<? extends PluginTypeInterface> pluginType, PluginInterface plugin ) {
+    lock.writeLock().lock();
+    try {
+      URLClassLoader ucl;
+      Set<PluginInterface> list = pluginMap.get( pluginType );
+      if ( list != null ) {
+        list.remove( plugin );
+      }
+
+      Map<PluginInterface, URLClassLoader> classLoaders = classLoaderMap.get( plugin.getPluginType() );
+      if ( classLoaders != null ) {
+        classLoaders.remove( plugin );
+      }
+
+      if ( !Utils.isEmpty( plugin.getClassLoaderGroup() ) ) {
+        // Straight away remove the class loader for the whole group...
+        //
+        ucl = classLoaderGroupsMap.remove( plugin.getClassLoaderGroup() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *ucl*. The resource is referenced at lines 184, 185. The resource is closed at line 189. There are other execution paths that do not contain closure statements, for example, when an else condition at line 184 evaluates to true. Ensure *ucl* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).